lib/TPP/LowerPacksAndUnpacks.cpp

```diff
@@ -0,0 +1,254 @@
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_LOWERPACKSANDUNPACKS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+namespace {
+
+// Tile by one up to the first dimension involved in packing.
+// 1) In case the pack as a consumer we can tile up to the first tiled dim.
+// 2) If the pack is a standalone op just make it 2d
+template <typename OpTy>
+static SmallVector<int64_t> getTileSizes(OpTy packingOp,
+                                         bool isConsumer = false) {
+  static_assert(llvm::is_one_of<OpTy, tensor::PackOp, tensor::UnPackOp>::value,
+                "applies to only pack or unpack operations");
+  SmallVector<int64_t> tiledDims = llvm::to_vector(packingOp.getInnerDimsPos());
+  assert(!tiledDims.empty());
+  if (isConsumer) {
+    int64_t upTo = *std::min_element(tiledDims.begin(), tiledDims.end());
+    return SmallVector<int64_t>(upTo, 1);
+  }
+  if (std::is_same<OpTy, tensor::PackOp>::value) {
+    int64_t upTo = packingOp.getDestType().getRank() - 2;
+    return SmallVector<int64_t>(upTo, 1);
+  }
+  return llvm::to_vector(packingOp.getStaticTiles());
+}
+
+// Tile`operation` using `tileSizes`, the loops are annotated with
+// `kLoopParallel` and later converted to scf.forall.
+static FailureOr<scf::SCFTilingResult> tileOp(RewriterBase &rewriter,
+                                              TilingInterface operation,
+                                              ArrayRef<int64_t> tileSizes) {
+  auto options = scf::SCFTilingOptions().setTileSizes(
+      getAsIndexOpFoldResult(rewriter.getContext(), tileSizes));
+  FailureOr<scf::SCFTilingResult> tilingResult =
+      scf::tileUsingSCFForOp(rewriter, operation, options);
+  if (failed(tilingResult))
+    return failure();
+  if (!tilingResult->loops.empty()) {
+    tilingResult->loops[0]->setAttr(
+        linalgx::utils::kLoopParallel,
+        rewriter.getStringAttr(linalgx::utils::kLoopRoot));
+  }
+  return tilingResult;
+}
+
+// Fuse producer and consumer pack. Standalone packs are tiled into 2d tiles.
+static void fuseOrTilePacks(RewriterBase &rewriter, FunctionOpInterface func) {
+  SmallVector<tensor::PackOp> consumerPackOps;
+  SmallVector<tensor::PackOp> otherPacks;
+  SmallVector<tensor::UnPackOp> unPacks;
+  func->walk<WalkOrder::PostOrder>([&](tensor::PackOp consumerPackOp) {
+    Value source = consumerPackOp.getSource();
+    tensor::PackOp producerPackOp =
+        dyn_cast_or_null<tensor::PackOp>(source.getDefiningOp());
+    if (producerPackOp)
+      consumerPackOps.push_back(consumerPackOp);
+    else
+      otherPacks.push_back(consumerPackOp);
+  });
+  func->walk<WalkOrder::PostOrder>(
+      [&](tensor::UnPackOp unpackOp) { unPacks.push_back(unpackOp); });
+
+  // Tile and fuse.
+  for (auto consumerPackOp : consumerPackOps) {
```

Comment:
Are you assuming here that the chained packs (consumers) are handled before the non-chained (producer) ones?

If you have a 3-chain, wouldn't the second be in the list before the third, breaking that rule?